pkg/kubelet/clustertrustbundle/clustertrustbundle_manager.go

```diff
@@ -0,0 +1,127 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package clustertrustbundle
+
+import (
+	"encoding/pem"
+	"fmt"
+	"sort"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	certinformersv1alpha1 "k8s.io/client-go/informers/certificates/v1alpha1"
+	certlistersv1alpha1 "k8s.io/client-go/listers/certificates/v1alpha1"
+	"k8s.io/client-go/tools/cache"
+)
+
+type Manager interface {
+	GetTrustAnchorsByName(name string) (string, error)
+	GetTrustAnchorsBySigner(signerName string, labelSelector metav1.LabelSelector) (string, error)
+}
+
+type InformerManager struct {
+	ctbInformer cache.SharedIndexInformer
+	ctbLister   certlistersv1alpha1.ClusterTrustBundleLister
+}
+
+func NewInformerManager(bundles certinformersv1alpha1.ClusterTrustBundleInformer) *InformerManager {
+	// We need to call Informer() before calling start on the shared informer
+	// factory, or the informer won't be registered to be started.
+	return &InformerManager{
+		ctbInformer: bundles.Informer(),
+		ctbLister:   bundles.Lister(),
+	}
+}
+
+func (m *InformerManager) GetTrustAnchorsByName(name string) (string, error) {
+	if !m.ctbInformer.HasSynced() {
+		return "", fmt.Errorf("ClusterTrustBundle informer has not yet synced")
+	}
+
+	ctb, err := m.ctbLister.Get(name)
+	if err != nil {
+		return "", fmt.Errorf("while getting ClusterTrustBundle: %w", err)
+	}
+
+	return ctb.Spec.TrustBundle, nil
+}
+
+func (m *InformerManager) GetTrustAnchorsBySigner(signerName string, labelSelector metav1.LabelSelector) (string, error) {
```

Comment:
this gets called per-volume, per-pod, every few seconds... we likely need to avoid doing the same work to return content for different pods with identical signerName/selector specs, and avoid doing the same work on every call if the selected bundle objects have not changed

doing the label selection and signerName filtering on each call is probably fine, but before we do the PEM decode/re-encode, we should probably check if we already have computed content for the set of input names+resourceVersions, and cache that for some short period of time (maybe a minute or something)

Reply:
could we simplify the cache key to `{signerName,labelSelector.String()}`, and remove all cache entries for a given signerName whenever the informer observes an add/update/delete of a CTB for that signer, since we expect the CTB API to be slow-moving?